### DISCUSSION TITLE
Repisitory의 findByXXX() 중복사용에 대해

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [BCSDLab/KOIN_API_V2#161] [에러 / 고민사항] [answered]
> 온보딩 과정에서 발안된 주제입니다.

Repisitory의 findById() 중복사용에 대해 고민입니다.
현재는 최소한의 단위로 기능을 구현하면서 각 서비스별로 Repository에서 호출하는 `repository.findByXXX()` 가 단 한번씩만 존재하여 아래와 같이 사용되고 있습니다.

```java
// 예시 Method
public UserTokenRefreshResponse refresh(UserTokenRefreshRequest request) {
      ...
      UserToken userToken = userTokenRepository.findById(Long.parseLong(userId))
          .orElseThrow(() -> new IllegalArgumentException("refresh token이 존재하지 않습니다. request: " + request));
  ...
}
```

나중에 하나의 서비스에서 동일한 findByXXX() 가 중복되어 사용되는 경우 `orElseThrow()...` 구문이 중복하여 발생하게 됩니다.
이에 두개 이상의 findByXXX()가 등장할 경우 private 메소드로 묶어서 사용하는것은 어떤지 의견을 묻고싶습니다!

개인적으로는 getByXXX라는 네이밍이 생각나네요

```java
private Object getByXXX(String XXX) {
    return someThingRepository.findByXXX(XXX)
              .orElseThrow(() -> new SomethingException.withDetail("detail message XXX: " + XXX);
}
```

다양한 의견 부탁드립니다 (동의도 좋아요~)

**Comment by songsunkook** [comment, marked as answer]
Service 단에서 private 메서드로 분리하는 것은 결국 여러 클래스 간의 코드 중복을 막지 못해서 근본적으로 의미가 적다고 생각합니다. 차라리 인터페이스의 default 메서드 기능을 활용해서 Repository 단에서 예외처리를 한 번에 수행하는 것은 어떨까요??

Repository 단에서 default 메서드를 도입할 시 얻을 수 있는 이점
- Optional 예외의 내부 처리
  - 호출자의 관심사 집중, 코드 간소화 가능
- 메서드명 간소화
  - 쿼리가 복잡해질 경우 메서드명이 길어져 가독성이 오히려 떨어지는 현상을 막을 수 있음

[참고 블로그](https://dkswnkk.tistory.com/730)

**Reply by Choi-JJunho** [reply]
오! default 메소드 좋네요~!
Wiki에 컨벤션으로 추가한 뒤 빠르게 도입해보는 것도 좋을 것 같네요 👍

- [Wiki - 코드 컨벤션](https://github.com/BCSDLab/KOIN_API_V2/wiki/%F0%9F%93%9A-%EC%BD%94%EB%93%9C-%EC%BB%A8%EB%B2%A4%EC%85%98)

**Reply by xhdtn8070** [reply]
abstract 클래스를 만들고 상속시키는 방법을 생각했었는데,
interface에 default 만드는 이 방법이 훨씬 좋아보이네요!
잘 배웠습니다~

**Reply by daheeParkk** [reply]
좋은 방법인 것 같습니다!
default 메서드를 이렇게 사용할 수 있다는 것을 배웠습니다!

**Comment by Choi-JJunho** [comment]
해당 부분에 대해 Wiki에 문서화 진행했습니다.

- [📚 코드 컨벤션 - 메서드 이름](https://github.com/BCSDLab/KOIN_API_V2/wiki/%F0%9F%93%9A-%EC%BD%94%EB%93%9C-%EC%BB%A8%EB%B2%A4%EC%85%98#%EB%A9%94%EC%84%9C%EB%93%9C-%EC%9D%B4%EB%A6%84)
